### ISSUE TITLE
Unable to split up a config file and use .get in child require

### ISSUE THREAD

**Issue by timkelty** [lorenwest/node-config#203]
I'm trying to split up my `config/default.js` file into chunks, but running into problems.
Here's a simplified version of what I'm trying to do:

## config/default.js

```js
module.exports = {
  paths: {
    src: './app/',
    dest: './public/',
  },
  webpack: require('./webpack'),
};
```

## config/webpack.js

```js
var config = require('config');
module.exports = {
    entry: {
      'head': [path.join(config.get('paths.src'), 'assets/js/head.js')],
      'tail': [path.join(config.get('paths.src'), 'assets/js/tail.js')]
    },
};
```

Which gets me:
> Error: Cannot parse config file: '/Users/timkelty/Sites/edwards/config/default.js': TypeError: Object #<Object> has no method 'get'

Is there a way to use config items from a child require like this?

**Comment by markstos** [comment]
There's a catch-22 here of wanting to use `config` within the `config` files.

The cycle can be broken by using the `config/defer` helper module, which allows you define a value as a function that is not evaluated until the end of the `config` resolution cycle. It receives a copy of the final config object, which can be used in the function.

In other words, the cycle is broken by allowing you /define/ functions that will /eventually/ use the `config` object, but don't do so during the main `config` building phase. 

There's an example use of the `defer` feature here:

https://github.com/lorenwest/node-config/wiki/Configuration-Files#javascript-module---js

**Comment by timkelty** [comment]
@markstos Thanks, that works great!

Am I correct that within the function, I would just have to use `cfg.paths.src` and not the `config.get` method?

**Comment by markstos** [comment]
Hmm, I believe so.   If the `get()` method doesn't work there, that's certainly the case. :)

I'm glad you found it useful @timkelty I wanted the same kind of feature, and this was one of my contributions to `node-config`.

**Comment by timkelty** [comment]
Exactly what I was looking for, thanks a ton @markstos!

**Comment by lorenwest** [comment]
Calling `require('config')` in any configuration file will get you into a loop.  There is no need for that in the sub-configuration file.  I think you'll find that removing `require('config')` in your sub-configuration file will resolve your issue without needing the defer.

**Comment by timkelty** [comment]
It is getting a little hairy when I try and combine this with environment merges...
E.g.:

## config/default.js

```js
module.exports = {
  paths: {
    src: './app/',
    dest: './public/',
  },
  webpack: require('./webpack'),
};
```

## config/webpack.js

```js
var defer = require('config/defer').deferConfig;
module.exports = defer(function(config) {
    entry: {
      'head': [path.join(config.paths.src, 'assets/js/head.js')],
      'tail': [path.join(config.paths.src, 'assets/js/tail.js')],
    },
});
```

## config/production.js

```js
module.exports = {
    webpack: {
        debug: false
    }
};
```

This doesn't seem to work. It seems that `config.webpack` is already set by time to gets to resolving the deferred, it doesn't do anything with it.

I expected to have to merge it myself in the `defer` callback, but it doesn't seem to even get there if `config.webpack` is already set.

**Comment by timkelty** [comment]
@lorenwest I tried that, but then `config` is not defined, and I get:
> Error: Cannot parse config file: '/Users/timkelty/Sites/edwards/config/default.js': ReferenceError: config is not defined

**Comment by lorenwest** [comment]
I see now - you're attempting to perform post-configuration work before configuration has completed.  I don't believe what you're attempting to do would work - regardless of it being in a config.js or a sub-config .js file. 

The way I would approach the issue is to move that code up into your application bootstrap.

I used to have heated arguments with a colleague about allowing code execution during configuration.  He told tales of doom if I allowed people to execute code in configuration files.

Well this isn't exactly doom, but it is what he was talking about.  Configuration files should essentially contain static information.  Any real work should be done at a higher level.
